openedx/core/djangoapps/content_libraries/api/containers.py

```diff
@@ -610,3 +614,186 @@ def publish_container_changes(container_key: LibraryContainerLocator, user_id: i
+                continue
+
+            if isinstance(item, Container):
+                metadata = ContainerMetadata.from_container(
+                    self.object_key.context_key,
+                    item,
+                )
+            else:
+                assert isinstance(item, ContainerMetadata)
+                metadata = item
+
+            if level == 'units':
+                self.units.append(metadata)
+            elif level == 'subsections':
+                self.subsections.append(metadata)
+            elif level == 'sections':
+                self.sections.append(metadata)
+            else:
+                raise TypeError(f"Invalid level: {level}")
+
+    @staticmethod
+    def parent_level(level: str | None) -> str | None:
+        """
+        Returns the name of the parent field above the given level,
+        or None if level is already the top level.
+        """
+        match level:
+            case "components":
+                return "units"
+            case "units":
+                return "subsections"
+            case "subsections":
+                return "sections"
+            case _:
+                return None
+
+    @staticmethod
+    def child_level(level: str | None) -> str | None:
+        """
+        Returns the name of the child field below the given level,
+        or None if level is already the lowest level.
+        """
+        match level:
+            case "sections":
+                return "subsections"
+            case "subsections":
+                return "units"
+            case "units":
+                return "components"
+            case _:
+                return None
+
+    @classmethod
+    def create_from_library_object_key(
+        cls,
+        object_key: LibraryUsageLocatorV2 | LibraryContainerLocator,
+    ):
+        """
+        Returns a LibraryObjectHierarchy populated from the library object represented by the given object_key.
+        """
+        root_items: list[Component] | list[Container]
+        root_level: str
+
+        if isinstance(object_key, LibraryUsageLocatorV2):
+            root_items = [get_component_from_usage_key(object_key)]
+            root_level = "components"
+
+        elif isinstance(object_key, LibraryContainerLocator):
+            root_items = [_get_container_from_key(object_key)]
+            root_level = f"{object_key.container_type}s"
+
+        else:
+            raise TypeError(f"Unexpected '{object_key}': must be LibraryUsageLocatorv2 or LibraryContainerLocator")
+
+        # Fill in root level of hierarchy
+        hierarchy = cls(object_key=object_key)
+        items = root_items
+        hierarchy.append(root_level, *items)
+        level: str | None = root_level
+
+        # Fill in hierarchy up through parents
+        while level := hierarchy.parent_level(level):
+            items = list(_get_containers_with_entities(items).all())
+            hierarchy.append(level, *items)
```

Comment:
No action needed: Via https://github.com/openedx/openedx-learning/pull/316 we will soon be adding a new `OutlineRoot` container type that's used for courses and which can have child containers of any type, including a mix. I don't think this code will need to support that (since they're only for courses not libraries), but thought I should mention it, especially if we ever move this "get hierarchy" functionality into learning core instead of just keeping it in content_libraries.

Reply:
We might have to move this into openedx-learning anyway to optimize the queries, so I'll keep this in mind.